File: bin/.common/logger.rb
# frozen_string_literal: true

require 'pastel'

# Centralized logging system for ZSH configuration scripts
# Provides consistent colored output and emoji indicators
class Logger
  def initialize
    @pastel = Pastel.new
    @color_enabled = true
  end

  # Core logging methods with emoji indicators
  def info(message)
    puts colorize("ℹ️  #{message}", :blue)
  end

  def success(message)
    puts colorize("✅ #{message}", :green)
  end

  def warning(message)
    puts colorize("⚠️  #{message}", :yellow)
  end

  def error(message)
    $stderr.puts colorize("❌ #{message}", :red)
  end

  def progress(message)
    puts colorize("🔄 #{message}", :cyan)
  end

  def section(message)
    puts colorize("🔧 #{message}", :magenta)
  end

  def debug(message)
    return unless ENV['DEBUG'] == '1'
    puts colorize("🐛 #{message}", :dim)
  end

  # Specialized logging methods
  def file_created(path)
    success("Created file: #{File.basename(path)}")
    debug("Path: #{path}")
  end

  def file_updated(path)
    info("📝 Updated: #{File.basename(path)}")
    debug("Path: #{path}")
  end

  def file_backed_up(path)
    info("💾 Backed up: #{File.basename(path)}")
    debug("Path: #{path}")
  end

  def install(package)
    success("📦 Installed: #{package}")
  end

  def clean(item)
    info("🧹 Cleaned: #{item}")
  end

  def update(item)
    info("🔄 Updated: #{item}")
  end

  # Platform-specific logging
  def brew(message)
    info("🍺 #{message}")
  end

  def git(message)
    info("🐙 #{message}")
  end

  def python(message)
    info("🐍 #{message}")
  end

  def ruby(message)
    info("💎 #{message}")
  end

  def macos(message)
    info("🍎 #{message}")
  end

  def linux(message)
    info("🐧 #{message}")
  end

  # Utility methods
  def separator
    puts colorize("━" * 50, :bright_black)
  end

  def complete(process_name)
    puts
    success("🎉 #{process_name} completed successfully!")
    puts
  end

  def banner(title)
    separator
    puts colorize("  #{title.upcase}", :bold)
    separator
  end

  # Configuration
  def verbose?
    ENV['VERBOSE'] == '1' || ENV['V'] == '1'
  end

  def disable_colors!
    @color_enabled = false
  end

  def enable_colors!
    @color_enabled = true
  end

  private

  def colorize(text, color)
    return text unless @color_enabled
    @pastel.send(color, text)
  end
end

# Global logger instance
$logger = Logger.new

# Convenience methods for global access
def log_info(message)
  $logger.info(message)
end

def log_success(message)
  $logger.success(message)
end

def log_warning(message)
  $logger.warning(message)
end

def log_error(message)
  $logger.error(message)
end

def log_progress(message)
  $logger.progress(message)
end

def log_section(message)
  $logger.section(message)
end

def log_debug(message)
  $logger.debug(message)
end

def log_file_created(path)
  $logger.file_created(path)
end

def log_file_updated(path)
  $logger.file_updated(path)
end

def log_file_backed_up(path)
  $logger.file_backed_up(path)
end

def log_install(package)
  $logger.install(package)
end

def log_clean(item)
  $logger.clean(item)
end

def log_update(item)
  $logger.update(item)
end

def log_brew(message)
  $logger.brew(message)
end

def log_git(message)
  $logger.git(message)
end

def log_python(message)
  $logger.python(message)
end

def log_ruby(message)
  $logger.ruby(message)
end

def log_macos(message)
  $logger.macos(message)
end

def log_linux(message)
  $logger.linux(message)
end

def log_separator
  $logger.separator
end

def log_complete(process_name)
  $logger.complete(process_name)
end

def log_banner(title)
  $logger.banner(title)
end

File: bin/bun-cli/src/core/utils/logger.ts
import type { Logger as LoggerInterface } from "../types";

/**
 * Logger implementation with emoji indicators and colors
 *
 * Simple, no dependencies, just ANSI escape codes
 */
export class Logger implements LoggerInterface {
  private verbose: boolean;
  private debug: boolean;

  constructor(params: { verbose?: boolean; debug?: boolean } = {}) {
    this.verbose = params.verbose ?? false;
    this.debug = params.debug ?? (process.env.DEBUG === "1");
  }

  info(message: string): void {
    console.log(`ℹ️  ${this.blue(message)}`);
  }

  success(message: string): void {
    console.log(`✅ ${this.green(message)}`);
  }

  warn(message: string): void {
    console.log(`⚠️  ${this.yellow(message)}`);
  }

  error(message: string): void {
    console.error(`❌ ${this.red(message)}`);
  }

  progress(message: string): void {
    console.log(`🔄 ${this.cyan(message)}`);
  }

  section(title: string): void {
    console.log(`\n🔧 ${this.magenta(title)}`);
  }

  banner(title: string): void {
    const line = "=".repeat(60);
    console.log(`\n${line}\n  ${title}\n${line}\n`);
  }

  debug(message: string): void {
    if (this.debug) {
      console.log(`🐛 ${this.dim(message)}`);
    }
  }

  // Color helpers
  private blue(s: string): string {
    return `\x1b[34m${s}\x1b[0m`;
  }

  private green(s: string): string {
    return `\x1b[32m${s}\x1b[0m`;
  }

  private yellow(s: string): string {
    return `\x1b[33m${s}\x1b[0m`;
  }

  private red(s: string): string {
    return `\x1b[31m${s}\x1b[0m`;
  }

  private cyan(s: string): string {
    return `\x1b[36m${s}\x1b[0m`;
  }

  private magenta(s: string): string {
    return `\x1b[35m${s}\x1b[0m`;
  }

  private dim(s: string): string {
    return `\x1b[2m${s}\x1b[0m`;
  }
}
